### PR TITLE
Add Action Cable testing

### LINKED ISSUE #1606: Work out testing story for ActionCable
We currently don't have one, so we should definitely work out what the story is.

### PR DESCRIPTION
**Pull request by palkan**
This PR migrates [`action-cable-testing` gem](https://github.com/palkan/action-cable-testing) RSpec functionality into `rspec-rails` (since core functionality has been merged into Rails 6).

You can check the existing gem documentation (RSpec-like) to see what's coming: https://relishapp.com/palkan/action-cable-testing/docs.

## What's inside:
- [x] **Added `@rails_post_6` tag for 🥒**
- [x] Add `have_broadcasted_to` / `broadcast_to` matcher (=> corresponds to https://github.com/rails/rails/pull/33659)
- [x] Add channels unit-testing support ( => corresponds to https://github.com/rails/rails/pull/33969)
- [x] Add connections unit-testing support (=> corresponds to https://github.com/rails/rails/pull/34845)
- [x] Generators (=> corresponds to https://github.com/rails/rails/pull/34933).

Closes #1606

### CODE REVIEW

**Comment by palkan** [comment]
Ok, travis doesn't run builds for drafts; falling back to good old `[WIP]`

**Comment by palkan** [comment]
@benoittgt @samphippen Hey folks, what should I do to make Travis show me the failures?)

**Comment by benoittgt** [comment]
Hello @palkan 

Sorry busy weekend. Maybe you can comment those lines?
https://github.com/rspec/rspec-rails/blob/master/.travis.yml#L210-L213

I will have a look on your PR tomorrow. Thanks for the hard work.

**Comment by samphippen** [comment]
@palkan please rebase this on top of the 4-0-dev branch, as this'll have to make it in to 4-0 at this point.

**Comment by palkan** [comment]
@samphippen 
>  please rebase this on top of the 4-0-dev branch

Done!

File: features/matchers/have_broadcasted_matcher.feature [hunk 1]
```diff
@@ -0,0 +1,151 @@
+Feature: have_broadcasted matcher
```
Comment (benoittgt) [inline features/matchers/have_broadcasted_matcher.feature]: Don't we need documentation for other matchers too? Documentation display only one matcher doc. https://relishapp.com/palkan/action-cable-testing/docs/matchers/have-broadcasted-matcher

Comment (palkan) [reply features/matchers/have_broadcasted_matcher.feature]: Yeah, I think it makes sense to add a _matchers_ doc. We already have some example group specific matchers there (like `render_template`), so it should be OK.

Update the gem's docs: https://relishapp.com/palkan/action-cable-testing/docs/matchers/have-stream-from-matcher

If it looks good I update the PR as well.

File: example_app_generator/generate_stuff.rb [hunk 2]
```diff
@@ -111,6 +111,12 @@ def using_source_path(path)
 rescue LoadError
 end
 
+begin
+  require 'action_cable'
+  generate('channel chat') if ::Rails::VERSION::STRING.to_f >= 6.0
+rescue LoadError
```
Comment (JonRowe) [inline example_app_generator/generate_stuff.rb]: When would this generate a load error when on Rails 6? Seems like the two gates could be combined?

File: features/channel_specs/channel_spec.feature [hunk 3]
```diff
@@ -0,0 +1,296 @@
+Feature: channel spec
```
Comment (JonRowe) [inline features/channel_specs/channel_spec.feature]: Should all these features be gated to rails 6?

File: features/matchers/have_broadcasted_matcher.feature (hunk 1 above)
Comment (JonRowe) [inline features/matchers/have_broadcasted_matcher.feature]: Should all these features be gated to rails 6?

File: lib/rspec/rails/example/channel_example_group.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    # Container module for channel spec functionality. It is only available if
+    # ActionCable has been loaded before it.
+    module ChannelExampleGroup
+      # This blank module is only necessary for YARD processing. It doesn't
+      # handle the conditional `defined?` check below very well.
+    end
+  end
+end
+
+if defined?(ActionCable) && ::Rails::VERSION::MAJOR > 5
```
Comment (JonRowe) [inline lib/rspec/rails/example/channel_example_group.rb]: Surely either would suffice?

File: features/matchers/have_broadcasted_matcher.feature (hunk 1 above)
Comment (benoittgt) [reply features/matchers/have_broadcasted_matcher.feature]: LGTM. Thanks

**Comment by palkan** [comment]
@benoittgt @JonRowe thanks for the feedback! All fixed.

@samphippen The build seems green: https://travis-ci.org/rspec/rspec-rails/builds/526676636 (Rails 4.2 specs are failing for non-related to this PR reasons, looks it's no longer supported?).
Also, to make the build pass the RuboCop check I had to freeze the version (`~> 0.67.0`), the new one breaks it (yeah, as always). That should be fixed in the base branch, I think.

**Comment by JonRowe** [comment]
@samphippen can you give me a final review for this, you're more familiar with action cable!

File: lib/rspec/rails/example/channel_example_group.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      # @private
+      module ClassMethods
+        # These blank modules are only necessary for YARD processing. It doesn't
+        # handle the conditional check below very well and reports undocumented objects.
+      end
+    end
+  end
+end
+
+if ::Rails::VERSION::MAJOR > 5
```
Comment (samphippen) [inline lib/rspec/rails/example/channel_example_group.rb]: WDYT about `>= 6`?

File: lib/rspec/rails/matchers/action_cable.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      #     }.to have_broadcasted_to("messages").at_least(2).times
+      #
+      #     expect {
+      #       ActionCable.server.broadcast "messages", text: 'Hi!'
+      #     }.to have_broadcasted_to("messages").at_most(:twice)
+      #
+      #     expect {
+      #       ActionCable.server.broadcast "messages", text: 'Hi!'
+      #     }.to have_broadcasted_to("messages").with(text: 'Hi!')
+
```
Comment (samphippen) [inline lib/rspec/rails/matchers/action_cable.rb]: you need to kill this blankline.

**Comment by samphippen** [comment]
@palkan BTW you should feel free to DM me on twitter if you need my attention, it is by far and away my fastest response time channel.

**Comment by samphippen** [comment]
LGTM, will merge on green!

**Comment by palkan** [comment]
Looks like we're blocked by RuboCop again)

image

@benoittgt @samphippen Is anyone going to fix this in `4-0-dev`?

**Comment by benoittgt** [comment]
@palkan We merged https://github.com/rspec/rspec-rails/pull/2126 this afternoon sorry. It should be good. 😊

**Comment by palkan** [comment]
@benoittgt My bad, I should have checked(

Anyway, we're green 🎉

**Comment by benoittgt** [comment]
Thanks a lot @palkan. Awesome job! 👏

**Comment by shyam04261996** [comment]
def speak(data)
    @chat = BxBlockChat::Chat.find(params[:id])
    message = BxBlockChat::ChatMessage.create!(account_id: current_user.id, chat_id: @chat.id, message: data["message"])
    if message && @chat.accounts
      @chat.accounts.each do |account|
        if account.id != current_user.id && account.status == "offline" && account.get_email_notifications
          ::BxBlockChat::ChatsMailer.with(
            to: account.email,
            chat: @chat.name
          ).new_message.deliver_now
        end
      end
    end
    ActionCable.server.broadcast build_room_id(@chat.id), message: message.message
  end

**Comment by pirj** [comment]
@shyam04261996 please use our help channels http://rspec.info/help/ or Stackoverflow.
We reserve the GutHub for feature requests/bug reports.

**Pull request merged**